src/hazmat/sieve.rs

```diff
@@ -77,45 +77,40 @@ impl<T: Integer> Sieve<T> {
 
         // If we are targeting safe primes, iterate over the corresponding
         // possible Germain primes (`n/2`), reducing the task to that with `safe_primes = false`.
-        let (max_bit_length, base) = if safe_primes {
+        let (max_bit_length, mut start) = if safe_primes {
             (max_bit_length - 1, start.wrapping_shr_vartime(1))
         } else {
             (max_bit_length, start)
         };
 
-        let mut base = base;
-
         // This is easier than making all the methods generic enough to handle these corner cases.
-        let produces_nothing = max_bit_length < base.bits_vartime() || max_bit_length < 2;
+        let produces_nothing = max_bit_length < start.bits_vartime() || max_bit_length < 2;
 
         // Add the exception to the produced candidates - the only one that doesn't fit
         // the general pattern of incrementing the base by 2.
         let mut starts_from_exception = false;
-        if base <= T::from(2u32) {
+        if start <= T::from(2u32) {
             starts_from_exception = true;
-            base = T::from(3u32);
+            start = T::from(3u32);
         } else {
-            // Adjust the base so that we hit odd numbers when incrementing it by 2.
-            base |= T::one();
+            // Adjust the start so that we hit odd numbers when incrementing it by 2.
+            start |= T::one();
         }
 
-        // Only calculate residues by primes up to and not including `base`,
-        // because when we only have the resiude,
-        // we cannot distinguish between a prime itself and a multiple of that prime.
-        let residues_len = if T::from(SMALL_PRIMES[SMALL_PRIMES.len() - 1]) >= base {
-            SMALL_PRIMES
-                .iter()
-                .enumerate()
-                .find(|(_i, p)| T::from(**p) >= base)
-                .map(|(i, _p)| i)
-                .unwrap_or(SMALL_PRIMES.len())
-        } else {
-            // This will be the majority of use cases
+        // Only calculate residues by primes up to and not including `start`, because when we only
+        // have the resiude, we cannot distinguish between a prime itself and a multiple of that
+        // prime.
+        let residues_len = if T::from(LAST_SMALL_PRIME) <= start {
             SMALL_PRIMES.len()
+        } else {
+            // `start` is smaller than the last prime in the list so casting `start` to a `u16` is
+            // safe. We need to find out how many residues we can use.
+            let start_u16 = start.as_ref()[0].0 as SmallPrime;
```

Comment:
Minor: since the type we're operating with here is called `SmallPrime`, the suffix should probably be `small` and not `u16`.